src/main/ts/core/Changelog.ts

```diff
@@ -0,0 +1,361 @@
+import * as commonmark from 'commonmark';
+import { DateTime } from 'luxon';
+import * as E from 'fp-ts/Either';
+import { pipe } from 'fp-ts/pipeable';
+import * as EitherUtils from '../utils/EitherUtils';
+import { parseVersionE, Version } from './Version';
+
+interface Offset {
+  start: number;
+  end: number;
+}
+
+interface Item extends Offset {
+  jira?: string;
+}
+
+interface Section extends Offset {
+  header: Offset;
+  list: Offset;
+  items: Item[];
+}
+
+type SectionNames = 'Added' | 'Improved' | 'Changed' | 'Deprecated' | 'Removed' | 'Fixed' | 'Security';
+
+interface ChangelogFragment {
+  sections: SectionNames[];
+  Added?: Section;
+  Improved?: Section;
+  Changed?: Section;
+  Deprecated?: Section;
+  Removed?: Section;
+  Fixed?: Section;
+  Security?: Section;
+}
+
+interface ReleaseMeta {
+  version: Version;
+  date: DateTime;
+}
+
+interface Release extends ChangelogFragment, Offset {
+  header: Offset;
+  meta?: ReleaseMeta;
+}
+
+export interface Changelog {
+  source: string;
+  releases: Release[];
+  preamble: Offset;
+  links: Offset;
+}
+
+interface TextLines {
+  text: string;
+  lines: number[];
+}
+
+interface Heading {
+  header: commonmark.Node;
+  preamble: commonmark.Node[];
+  subheadings: Heading[];
+}
+
+interface Top {
+  header?: undefined;
```

Comment:
Should this still be in there?

Reply:
Yep. I wanted to process Top and Heading with the same bit of code and if I don't have the property for header than typescript won't let me access it because it could be missing. My original version used Heading for both but this let me have better types.

Reply:
Added a comment to the code explaining why it is there.